### PR TITLE
TW-161: add preview file when user tap on file

### PR DESCRIPTION
**Pull request by sherlockvn**
https://github.com/linagora/twake-on-matrix/assets/43041967/f9954825-7a45-4c86-ba0d-53925944681d

https://github.com/linagora/twake-on-matrix/assets/43041967/c6c05ab4-1eaa-4e98-a52a-3da9b641ec97

### CODE REVIEW

File: lib/pages/chat/chat.dart [hunk 1]
```diff
@@ -74,6 +84,8 @@ class ChatController extends State<Chat> {
 
   final numberSelectedImagesNotifier = ValueNotifier<int>(0);
 
+  final StreamController<Either<Failure, Success>> streamController = StreamController();
```
Comment (hoangdat) [inline lib/pages/chat/chat.dart:87]: we need this?

File: lib/utils/voip/permission_service.dart
Comment (hoangdat) [inline lib/utils/voip/permission_service.dart:1]: srr, but why put this utils inside `voip`?

File: lib/utils/voip/permission_service.dart [hunk 2]
```diff
@@ -48,6 +48,19 @@ class PermissionHandlerService {
...
+  Future<bool> get hasStoragePermission async {
+    return await Permission.storage.isGranted;
+  }
+
+  Future<void> requestStoragePermission() async {
+    final status = await Permission.storage.status;
+    if (status.isDenied) {
+      Permission.storage.request();
+    } else if (status.isPermanentlyDenied) {
+      goToSettingsForPermissionActions();
```
Comment (hoangdat) [inline lib/utils/voip/permission_service.dart:60]: go directly? no inform user?

File: lib/pages/chat/events/message_download_content.dart [hunk 3]
```diff
@@ -54,7 +60,7 @@ class MessageDownloadContent extends StatelessWidget {
             child: Row(
               children: [
                 Text(
-                  filetype,
+                  filetype ?? 'UNKNOWN',
```
Comment (hoangdat) [inline lib/pages/chat/events/message_download_content.dart:63]: why is `UNKNOW`? how about other value?

Comment (hoangdat) [reply lib/pages/chat/events/message_download_content.dart:63]: IMO, we have the default `mimeType`, not a `UNKNOWN`, how about `application/octet-stream`?

File: lib/pages/chat/chat.dart [hunk 4]
```diff
@@ -444,6 +456,60 @@ class ChatController extends State<Chat> {
...
+            if (success is DownloadFileForPreviewSuccess) {
+              _openDownloadedFileForPreview(downloadFileForPreviewResponse: success.downloadFileForPreviewResponse);
+              Navigator.of(context).pop();
+            } else if (success is DownloadFileForPreviewLoading) {
+              showDialog(
+                context: context,
+                useRootNavigator: false,
+                builder: (BuildContext context) {
+                  return const Center(
+                    child: CircularProgressIndicator(),
```
Comment (hoangdat) [inline lib/pages/chat/chat.dart:486]: can you demo this case?

File: lib/utils/voip/permission_service.dart
Comment (sherlockvn) [reply lib/utils/voip/permission_service.dart:1]: yeah, not me. it's confusing me too, let me change it

File: lib/di/send_image/send_image_di.dart
Comment (nqhhdev) [inline lib/di/send_image/send_image_di.dart:1]: Should change `send_images/send_image_di.dart` WDYT?

File: lib/domain/usecase/download_file_for_preview_interactor.dart [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+import 'package:fluffychat/app_state/failure.dart';
+import 'package:fluffychat/app_state/success.dart';
+import 'package:fluffychat/domain/app_state/preview_file/download_file_for_preview_failure.dart';
+import 'package:fluffychat/domain/app_state/preview_file/download_file_for_preview_loading.dart';
+import 'package:fluffychat/domain/app_state/preview_file/download_file_for_preview_success.dart';
+import 'package:fluffychat/domain/model/download_file/download_file_for_preview_response.dart';
+import 'package:fluffychat/utils/matrix_sdk_extensions/event_extension.dart';
+import 'package:matrix/matrix.dart';
+import 'package:mime/mime.dart';
+import 'package:path_provider/path_provider.dart';
```
Comment (hoangdat) [inline lib/domain/usecase/download_file_for_preview_interactor.dart:15]: Please move it out of domain

File: lib/utils/matrix_sdk_extensions/event_extension.dart [hunk 6]
```diff
@@ -8,20 +8,33 @@ import 'package:fluffychat/utils/size_string.dart';
...
+  }
+
+  String? get fileType {
+    return (filename.contains('.')
+        ? filename.split('.').last.toUpperCase()
+        : content
+              .tryGetMap<String, dynamic>('info')
+              ?.tryGet<String>('mimetype')
+              ?.toUpperCase());
+  }
```
Comment (nqhhdev) [inline lib/utils/matrix_sdk_extensions/event_extension.dart:34]: indent

File: lib/utils/permission_dialog.dart [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    return Material(
+      color: Colors.transparent,
+      child: InkWell(
+        borderRadius: BorderRadius.circular(100.0),
+        onTap: onPressed, 
+        child: Padding(
+          padding: const EdgeInsets.symmetric(horizontal: 12.0, vertical: 10.0),
+          child: Text(text, style: Theme.of(context).textTheme.labelLarge?.copyWith(
+            color: Theme.of(context).colorScheme.primary
+          ),),
```
Comment (nqhhdev) [inline lib/utils/permission_dialog.dart:126]:
```suggestion
          )),
```

File: lib/pages/chat/chat.dart [hunk 8]
```diff
@@ -444,6 +458,102 @@ class ChatController extends State<Chat> {
...
+    final openResults = await OpenFile.open(
+      downloadFileForPreviewResponse.filePath,
+      type: mimeType,
+      uti: DocumentUti(SupportedPreviewFileTypes.iOSSupportedTypes[mimeType]).value
+    );
+    Logs().d('ChatController:_openDownloadedFileForPreview(): ${openResults.message}');
+    
+    if (openResults.type != ResultType.done) {
+      await Share.shareXFiles([XFile(downloadFileForPreviewResponse.filePath)]);
+      return ;
```
Comment (nqhhdev) [inline lib/pages/chat/chat.dart:553]:
```suggestion
      return;
```

**Review by hoangdat — APPROVED** [review]

**Pull request merged**